### ISSUE TITLE
Null pointer exceptions in GeoIP processor after update to 4.3

### ISSUE THREAD

**Issue by H2Cyber** [Graylog2/graylog2-server#12788]
## Current Behavior
Following an upgrade to Graylog 4.3.1, my server.log in now flooded with the following Warnings :

```
2022-06-01T23:23:58.280+01:00 WARN  [ProcessBufferProcessor] Unable to process message <87c84c80-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:58.666+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88030b91-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:58.667+01:00 WARN  [ProcessBufferProcessor] Unable to process message <880359b1-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:59.761+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88aa2107-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:59.769+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88ab8090-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:23:59.778+01:00 WARN  [ProcessBufferProcessor] Unable to process message <88acb916-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:00.390+01:00 WARN  [ProcessBufferProcessor] Unable to process message <890a1b66-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:00.682+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8936a992-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:00.949+01:00 WARN  [ProcessBufferProcessor] Unable to process message <895f6741-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:01.286+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8992d351-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:01.326+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8998eddd-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:01.727+01:00 WARN  [ProcessBufferProcessor] Unable to process message <89d61de5-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:02.058+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8a08c6a3-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerException
2022-06-01T23:24:03.330+01:00 WARN  [ProcessBufferProcessor] Unable to process message <8acade21-e1f9-11ec-8ea4-005056afb272>: java.lang.NullPointerExceptio
```

What can I do to understand what is causing these warnings and how to stop it ?

## Your Environment

* Graylog Version: 4.3
* Java Version: 8u312
* Elasticsearch Version: 7.10.2
* MongoDB Version: 4.0.28
* Operating System: Ubuntu Server 20.04

**Comment by HungryHowies** [comment]
I have just ran into the same issue after reading this post. I updated Graylog 4.3.0 to 4.3.1  and log spam
```
2022-06-01T21:58:26.098-05:00 WARN  [ProcessBufferProcessor] Unable to process message <df5d73f0-e21f-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T21:58:26.109-05:00 WARN  [ProcessBufferProcessor] Unable to process message <df5f21a3-e21f-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T21:58:29.306-05:00 WARN  [ProcessBufferProcessor] Unable to process message <e1471b81-e21f-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T21:59:20.529-05:00 WARN  [ProcessBufferProcessor] Unable to process message <ffcef6e3-e21f-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:01:39.382-05:00 WARN  [ProcessBufferProcessor] Unable to process message <52921920-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException.
2022-06-01T22:01:44.831-05:00 WARN  [ProcessBufferProcessor] Unable to process message <55d1b3c0-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:01:47.640-05:00 WARN  [ProcessBufferProcessor] Unable to process message <577e7960-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:02:20.663-05:00 WARN  [ProcessBufferProcessor] Unable to process message <6b2d1430-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:02:41.861-05:00 WARN  [ProcessBufferProcessor] Unable to process message <77cff134-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:03:01.786-05:00 WARN  [ProcessBufferProcessor] Unable to process message <83aff360-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException.
2022-06-01T22:03:45.963-05:00 WARN  [ProcessBufferProcessor] Unable to process message <9e052190-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:04:22.063-05:00 WARN  [ProcessBufferProcessor] Unable to process message <b3898dd0-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:04:22.130-05:00 WARN  [ProcessBufferProcessor] Unable to process message <b39378e0-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:05:03.945-05:00 WARN  [ProcessBufferProcessor] Unable to process message <cc801660-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException.
2022-06-01T22:06:24.515-05:00 WARN  [ProcessBufferProcessor] Unable to process message <fc84b874-e220-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:07:10.308-05:00 WARN  [ProcessBufferProcessor] Unable to process message <17d16600-e221-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
2022-06-01T22:07:37.267-05:00 WARN  [ProcessBufferProcessor] Unable to process message <27e35210-e221-11ec-a1b2-00155d601d11>: java.lang.NullPointerException
```

**Comment by bernd** [comment]
@H2Cyber Did you also update from 4.3.0 to 4.3.1?

@H2Cyber @HungryHowies Can you please enable `DEBUG` logging for the `ProcessBufferProcessor` class, so we get a stack trace? You can enable the `DEBUG` level at runtime with the following API call:

```
curl -X PUT -H x-requested-by:curl -u user:password https://your-graylog/api/cluster/system/loggers/org.graylog2.shared.buffers.processors.ProcessBufferProcessor/level/DEBUG
```

Thank you!

**Comment by H2Cyber** [comment]
@bernd in my case it was an upgrade from 4.2.9 to 4.3.1 using DEB packages.

**Comment by H2Cyber** [comment]
@bernd for how long should we enable `DEBUG` logging for the `ProcessBufferProcessor` class ? How do we disable it once done ?

**Comment by bernd** [comment]
> @bernd for how long should we enable `DEBUG` logging for the `ProcessBufferProcessor` class ? How do we disable it once done ?

We need a few `Unable to process message ...` log messages, including the stack trace. You can reset the log level to `INFO` by executing the same curl command again but using `INFO` instead of `DEBUG` in the URL.

**Comment by H2Cyber** [comment]
@bernd I did enable `DEBUG` logging for the `ProcessBufferProcessor` class using the provided command, but no stack trace appeared in the server.log file, just the continuous flood of java.lang.NullPointerException warnings. 

@HungryHowies any luck for you to get a stack trace ?

**Comment by bernd** [comment]
> @bernd I did enable `DEBUG` logging for the `ProcessBufferProcessor` class using the provided command, but no stack trace appeared in the `server.log` file, just the continuous flood of java.lang.NullPointerException warnings.
> 

@H2Cyber I updated the command. I missed the `-H Accept:application/json` parameter. Can you try the command again? Thank you!

**Comment by H2Cyber** [comment]
@bernd : no luck with the new command as well, but I suspect it may have to do with my password which contains special characters (explanation marks). 

Can I instead enable Debug logging using the GUI for `Subsystem: Graylog` under `System > Logging` ?

**Comment by bernd** [comment]
> @bernd : no luck with the new command as well, but I suspect it may have to do with my password which contains special characters (explanation marks).
> 
> Can I instead enable Debug logging using the GUI for `Subsystem: Graylog` under `System > Logging` ?

@H2Cyber Unfortunately there is no way to enable debugging for that single class. If it's your password, you can try to use single quotes around the value: `curl -u 'user:password'`. Also add the `-v` option to curl, that will give you more verbose output. Thank you!

**Comment by H2Cyber** [comment]
@bernd here is the output, if you need the message details I can share them but only privately :

```
2022-06-02T10:35:33.121+01:00 WARN  [ProcessBufferProcessor] Unable to process message <5960f003-e257-11ec-82a5-005056afb272>:
java.lang.NullPointerException: null
        at org.graylog.plugins.map.geoip.MaxMindIpLocationResolver.doGetGeoIpData(MaxMindIpLocationResolver.java:57) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolver.getGeoIpData(GeoIpResolver.java:64) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.addGeoIpDataIfPresent(GeoIpResolverEngine.java:111) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.filter(GeoIpResolverEngine.java:102) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.processor.GeoIpProcessor.process(GeoIpProcessor.java:94) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.handleMessage(ProcessBufferProcessor.java:158) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.dispatchMessage(ProcessBufferProcessor.java:128) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:98) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:49) [graylog.jar:?]
        at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
		
2022-06-02T10:35:33.290+01:00 WARN  [ProcessBufferProcessor] Unable to process message <597ab991-e257-11ec-82a5-005056afb272>:
java.lang.NullPointerException: null
        at org.graylog.plugins.map.geoip.MaxMindIpLocationResolver.doGetGeoIpData(MaxMindIpLocationResolver.java:57) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolver.getGeoIpData(GeoIpResolver.java:64) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.addGeoIpDataIfPresent(GeoIpResolverEngine.java:111) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.filter(GeoIpResolverEngine.java:102) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.processor.GeoIpProcessor.process(GeoIpProcessor.java:94) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.handleMessage(ProcessBufferProcessor.java:158) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.dispatchMessage(ProcessBufferProcessor.java:128) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:98) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:49) [graylog.jar:?]
        at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
		
2022-06-02T10:35:33.425+01:00 WARN  [ProcessBufferProcessor] Unable to process message <598f530a-e257-11ec-82a5-005056afb272>:
java.lang.NullPointerException: null
        at org.graylog.plugins.map.geoip.MaxMindIpLocationResolver.doGetGeoIpData(MaxMindIpLocationResolver.java:57) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolver.getGeoIpData(GeoIpResolver.java:64) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.addGeoIpDataIfPresent(GeoIpResolverEngine.java:111) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.filter(GeoIpResolverEngine.java:102) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.processor.GeoIpProcessor.process(GeoIpProcessor.java:94) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.handleMessage(ProcessBufferProcessor.java:158) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.dispatchMessage(ProcessBufferProcessor.java:128) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:98) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:49) [graylog.jar:?]
        at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
		
2022-06-02T10:35:33.429+01:00 WARN  [ProcessBufferProcessor] Unable to process message <598fc836-e257-11ec-82a5-005056afb272>:
java.lang.NullPointerException: null
        at org.graylog.plugins.map.geoip.MaxMindIpLocationResolver.doGetGeoIpData(MaxMindIpLocationResolver.java:57) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolver.getGeoIpData(GeoIpResolver.java:64) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.addGeoIpDataIfPresent(GeoIpResolverEngine.java:111) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.GeoIpResolverEngine.filter(GeoIpResolverEngine.java:102) ~[graylog.jar:?]
        at org.graylog.plugins.map.geoip.processor.GeoIpProcessor.process(GeoIpProcessor.java:94) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.handleMessage(ProcessBufferProcessor.java:158) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.dispatchMessage(ProcessBufferProcessor.java:128) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:98) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:49) [graylog.jar:?]
        at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
```

**Comment by bernd** [comment]
@H2Cyber Thank you, that helps a lot! There seems to be an issue with the GeoIP processor. We will investigate!

As a workaround, you can disable the GeoIP processor in the "Message Processors Configuration" on the "System / Overview" page:
![image](https://user-images.githubusercontent.com/461/171605824-4ec817af-74f6-49d0-a7ac-3d04ff98bf9c.png)

**Comment by HungryHowies** [comment]
Hello,

Sorry I'm late,
@H2Cyber  Yes, I was able to trace, but you  got to it before I could.
@bernd  Disabling  the GeoIP processor did resolve this issue  for me.

**Comment by danotorrey** [comment]
@HungryHowies @H2Cyber Thanks again for the quick report of the issue. We have drafted a fix for this, and we are now verifying that we have all of the bases covered with more testing. 

The issue is apparently caused by some IP addresses not returning a location from the MaxMind DB. We tried to reproduce that specific condition in a running Graylog cluster, but could not yet. From your descriptions, and large output of errors to the server log, it looks like the issue is happening quite a lot. Do you know if any specific IP addresses or ranges were triggering the issue, or did it seem to happen for all IPs? Are you able to tell us which version of the MaxMind DB you are using? Is it recent or older? Also, are you able to provide a screen shot of the System > Configuration > Geo-Location Processor config? Also, are the paths there to the DB field correct?

Thanks in advance for the additional info. We're just making sure that we understand all possible details around the issue to make sure we have the right fix in place.

**Comment by H2Cyber** [comment]
Thanks @danotorrey for the follow up. 

- Which version of the MaxMind DB you are using? Is it recent or older?
I was using a version that dates back to January 2022.

- Can you provide a screen shot of the System > Configuration > Geo-Location Processor config? Also, are the paths there to the DB field correct?
Yes the path is correct and it used to work flawlessly in GL 4.2.9 and older. Here is a screenshot :  
![Maxmind](https://user-images.githubusercontent.com/764189/171958624-0d99eafa-ca4c-4584-a538-c4b03cb2e0b8.png)

- What is approx message throughput rate in Graylog? Just trying to understand the rate of Geo lookups.
We're talking 5000+ events per second in my case. 

I have tried to update the MaxMind database with the latest GeoLite2-City file currently available from their website. Upon updating the  Graylog configuration to use the newest GeoLite2-City mmdb, I got the following Warnings in server.log :  
```
2022-06-03T22:42:58.125+01:00 WARN  [MaxMindIpResolver] Error creating DatabaseReader for 'MaxMindIpAsnResolver' with config file ''
2022-06-03T22:42:58.126+01:00 WARN  [MaxMindIpResolver] Error creating DatabaseReader for 'MaxMindIpAsnResolver' with config file ''
2022-06-03T22:42:58.126+01:00 WARN  [MaxMindIpResolver] Error creating DatabaseReader for 'MaxMindIpAsnResolver' with config file ''
2022-06-03T22:42:58.126+01:00 WARN  [MaxMindIpResolver] Error creating DatabaseReader for 'MaxMindIpAsnResolver' with config file ''
```

With the updated GeoLite2-City mmdb file in use, The Warning flood of `java.lang.NullPointerException` seems to have disappeared, and the geolocalisation seems to be working fine again. However, I noticed that the field names for geo-located-IPs have changed from :
```
src_city_name
src_country_code
src_geolocation
```

to :
```
src_geo_city
src_geo_coordinates
src_geo_country
src_geo_country_iso
src_geo_name
src_geo_region
src_geo_timezone
```
So to sum up, the issue is for me resolved followingan update of the GeoLite2-City file, however I will have to update all of my views , alerts and dashboards to use the new fields for geolocation.

**Comment by H2Cyber** [comment]
Update : in my case and even with the updated MaxMind file, the `java.lang.NullPointerException` are back in `server.log`.

**Comment by danotorrey** [comment]
@H2Cyber Thanks for the IPs that trigger the issue. That helped us test and verify our fix. 

@HungryHowies Thanks for the details. As far as I can tell, the issue happens for specific IP addresses (for example those in the comment above). The error is likely happening when one of those IPs comes through. Nevertheless, this was confirmed to be happening due to a bug, and the fix will be included in the next Graylog `4.3.2` bug fix release.
